File: code-generator/src/main/java/io/github/potjerodekool/openapi/ApiConfiguration.java
package io.github.potjerodekool.openapi;

import java.io.File;
import java.util.Map;

public record ApiConfiguration(File apiFile,
                               File schemasDir,
                               File pathsDir,
                               String basePackageName,
                               boolean generateApiDefinitions,
                               boolean generateModels,
                               Map<String, String> controllers) {

    public ApiConfiguration(File apiFile,
                            String basePackageName,
                            boolean generateApiDefinitions,
                            boolean generateModels,
                            Map<String, String> controllers) {
        this(apiFile, createRelativeFile(apiFile, "schemas"), createRelativeFile(apiFile, "paths"), basePackageName, generateApiDefinitions, generateModels, controllers);
    }

    private static File createRelativeFile(final File file,
                                           final String name) {
        return new File(file.getParentFile(), name);
    }

    public String modelPackageName() {
        return basePackageName + ".model";
    }

    public ApiConfiguration withControllers(Map<String, String> controllers) {
        return new ApiConfiguration(
                apiFile,
                basePackageName,
                generateApiDefinitions,
                generateModels,
                controllers
        );
    }
}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/ApiConfiguration.kt
package io.github.potjerodekool.openapi.common

import java.io.File

data class ApiConfiguration(
    val apiFile: File,
    val schemasDir: File,
    val pathsDir: File,
    val basePackageName: String,
    val isGenerateApiDefinitions: Boolean,
    val generateApiImplementations: Boolean,
    val generateModels: Boolean
) {
    constructor(
        apiFile: File,
        basePackageName: String,
        generateApiDefinitions: Boolean,
        generateApiImplementations: Boolean,
        generateModels: Boolean
    ) : this(
        apiFile,
        createRelativeFile(apiFile, "schemas"),
        createRelativeFile(apiFile, "paths"),
        basePackageName,
        generateApiDefinitions,
        generateApiImplementations,
        generateModels
    )

    fun modelPackageName(): String {
        return "$basePackageName.model"
    }

    companion object {
        private fun createRelativeFile(
            file: File,
            name: String
        ): File {
            return File(file.parentFile, name)
        }
    }
}

